### PR TITLE
Merge NServiceBus.Serverless into AzureFunctions

### LINKED ISSUE #6: Align public namespace with naming rules
https://github.com/Particular/PlatformArchitecture/blob/master/principles/cross-cutting-concepts.md#downstream-components-rules

* For public types relevant to business developers, we use the `NServiceBus` namespace to make them more discoverable and to avoid extra using statements
* For internal types, we use the root [`{Component}`](#component-naming-rules) namespace. This allows for unique identification of those types in logs and stack traces.
   - Examples: `NServiceBus.Gateway`, `NServiceBus.Persistence.AzureStorage` etc.
* For public types designed for extensibility, we use the root [`{Component}`](#component-naming-rules) namespace. This hides types irrelevant for business developers while making them discoverable for developers extending NServiceBus.

### PR DESCRIPTION
**Pull request by SeanFeldman**
fixes #6

### CODE REVIEW

File: src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt [hunk 1]
```diff
@@ -6,18 +6,37 @@ namespace NServiceBus.AzureFunctions
         public Microsoft.Azure.WebJobs.ExecutionContext ExecutionContext { get; }
         public Microsoft.Extensions.Logging.ILogger Logger { get; }
     }
+    public abstract class ServerlessEndpointConfiguration
```
Comment (timbussmann) [inline src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:9]: do those types need to be public anymore?

File: src/Shared/Serverless/ServerlessEndpoint.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Transport;
+
+    /// <summary>
+    /// An NServiceBus endpoint which does not receive messages automatically but only handles messages explicitly passed to it
+    /// by the caller.
+    /// Instances of <see cref="ServerlessEndpoint{TExecutionContext, TConfiguration}" /> can be cached and are thread-safe.
+    /// </summary>
+    public abstract class ServerlessEndpoint<TExecutionContext, TConfiguration>
```
Comment (timbussmann) [inline src/Shared/Serverless/ServerlessEndpoint.cs:13]: should those be private?

File: src/Shared/Serverless/ServerlessEndpointConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        public EndpointConfiguration AdvancedConfiguration => EndpointConfiguration;
+
+        /// <summary>
+        /// Define a transport to be used when sending and publishing messages.
+        /// </summary>
+        protected TransportExtensions<TTransport> UseTransport<TTransport>()
+            where TTransport : TransportDefinition, new()
+        {
+            var serverlessTransport = EndpointConfiguration.UseTransport<ServerlessTransport<TTransport>>();
+            //TODO improve
```
Comment (timbussmann) [inline src/Shared/Serverless/ServerlessEndpointConfiguration.cs:43]: let's remove the comment or address this.

File: src/Shared/Serverless/TransportWrapper/PipelineInvoker.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    class PipelineInvoker : IPushMessages
+    {
+        Task IPushMessages.Init(Func<MessageContext, Task> onMessage, Func<ErrorContext, Task<ErrorHandleResult>> onError, CriticalError criticalError, PushSettings settings)
+        {
+            if (this.onMessage == null)
+            {
+                // The core ReceiveComponent calls TransportInfrastructure.MessagePumpFactory() multiple times
+                // the first invocation is for the main pipeline, ignore all other pipelines as we don't want to manually invoke them.
+                this.onMessage = onMessage;
```
Comment (timbussmann) [inline src/Shared/Serverless/TransportWrapper/PipelineInvoker.cs:15]: we should raise this hack to core to be addressed by a proper API

File: src/StorageQueues.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt [hunk 5]
```diff
@@ -6,18 +6,37 @@ namespace NServiceBus.AzureFunctions
         public Microsoft.Azure.WebJobs.ExecutionContext ExecutionContext { get; }
         public Microsoft.Extensions.Logging.ILogger Logger { get; }
     }
+    public abstract class ServerlessEndpointConfiguration
```
Comment (timbussmann) [inline src/StorageQueues.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:9]: same here, we should probably review the public types

File: src/Shared/Serverless/ServerlessEndpoint.cs (hunk 2 above)
Comment (SeanFeldman) [reply src/Shared/Serverless/ServerlessEndpoint.cs:13]: `FunctionEndpoint` is a public type and implements `ServerlessEndpoint`. Therefore base type has to have the same accessibility.

> Error	CS0060	Inconsistent accessibility: base class > 'ServerlessEndpoint<FunctionExecutionContext, StorageQueueTriggeredEndpointConfiguration>' is less accessible than class 'FunctionEndpoint'	NServiceBus.AzureFunctions.StorageQueues

File: src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt (hunk 1 above)
Comment (SeanFeldman) [reply src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:9]: Has to stay public. We expose certain extension points such as `AdvancedConfiguration()`, `UseSerialization<T>`, and `DoNotSendMessagesToErrorQueue()` to the users.

File: src/StorageQueues.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt (hunk 5 above)
Comment (SeanFeldman) [reply src/StorageQueues.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:9]: We can't for the same reason as here https://github.com/Particular/NServiceBus.AzureFunctions/pull/41#discussion_r434096996

**Pull request merged**